cpp/src/join/joining.cu

```diff
@@ -242,149 +210,175 @@ gdf_error join_call( int num_cols, gdf_column **leftcol, gdf_column **rightcol,
+        cudf::table const& rjoincols,
+        cudf::table const& left_cols, 
+        cudf::table const& right_cols,
+        cudf::table const& join_cols,
+        cudf::table const& merging_join_cols,
         gdf_column * left_indices,
         gdf_column * right_indices) {
 
 
-  PUSH_RANGE("LIBGDF_JOIN_OUTPUT", JOIN_COLOR);
+    PUSH_RANGE("LIBGDF_JOIN_OUTPUT", JOIN_COLOR);
     //create left and right input table with columns not joined on
+    int num_left_cols = left_cols.num_columns();
+    int num_right_cols = right_cols.num_columns();
+    int num_cols_to_join = (join_cols.num_columns() > 0 ? join_cols.get_column(0)->size: 0);
+    int num_merging_join_cols = (merging_join_cols.num_columns() > 0 ? merging_join_cols.get_column(0)->size: 0);
+
+    std::vector<int> left_j_cols (num_cols_to_join);
+    std::vector<int> l_common_name_join_ind (num_merging_join_cols);
+    std::vector<int> r_common_name_join_ind (num_merging_join_cols);
+
+    if (num_cols_to_join > 0)
+    {
+        CUDA_TRY (cudaMemcpy(left_j_cols.data(), join_cols.get_column(0)->data, 
+                             cudf::size_of(join_cols.get_column(0)->dtype)*num_cols_to_join , cudaMemcpyDeviceToHost));
+    }
+    if (num_merging_join_cols > 0)
+    {
+        CUDA_TRY (cudaMemcpy(l_common_name_join_ind.data(), merging_join_cols.get_column(0)->data,
+                               cudf::size_of(merging_join_cols.get_column(0)->dtype)*num_merging_join_cols , cudaMemcpyDeviceToHost));
+        CUDA_TRY (cudaMemcpy(r_common_name_join_ind.data(), merging_join_cols.get_column(1)->data,
+                               cudf::size_of(merging_join_cols.get_column(1)->dtype)*num_merging_join_cols , cudaMemcpyDeviceToHost));
+    }
+    
+    gdf_size_type join_size = left_indices->size;
+    std::vector <gdf_dtype> rdtypes;
+    std::vector <gdf_dtype_extra_info> rdtype_infos;
+
     std::vector<gdf_column*> lnonjoincol;
     std::vector<gdf_column*> rnonjoincol;
-    std::set<int> l_join_indices, r_join_indices;
-    for (int i = 0; i < num_cols_to_join; ++i) {
-        l_join_indices.insert(left_join_cols[i]);
-        r_join_indices.insert(right_join_cols[i]);
-    }
     for (int i = 0; i < num_left_cols; ++i) {
-        if (l_join_indices.find(i) == l_join_indices.end()) {
-            lnonjoincol.push_back(left_cols[i]);
+        if (std::find(l_common_name_join_ind.begin(), l_common_name_join_ind.end(), i)
+            == l_common_name_join_ind.end()) {
+            lnonjoincol.push_back(const_cast<gdf_column*>(left_cols.get_column(i)));
         }
     }
     for (int i = 0; i < num_right_cols; ++i) {
-        if (r_join_indices.find(i) == r_join_indices.end()) {
-            rnonjoincol.push_back(right_cols[i]);
+        if (std::find(r_common_name_join_ind.begin(), r_common_name_join_ind.end(), i) 
+            == r_common_name_join_ind.end()) {
+            rnonjoincol.push_back(const_cast<gdf_column*>(right_cols.get_column(i)));
```

Comment:
Same as above: set difference followed by gathering up the columns, types, and dtype_infos.